### PR TITLE
item::reload: fix use-after-free bug with ammo pointer

### PR DESCRIPTION
**Pull request by tycho**
Found with Address Sanitizer (-fsanitize=address).

### CODE REVIEW

**Comment by tycho** [comment]
For what it's worth, this was the only ASAN issue I discovered when running the full test suite and running through the tutorial. So.. nice job, dev team!

File: src/item.cpp [hunk 1]
```diff
@@ -4616,6 +4616,7 @@ bool item::reload( player &u, item_location loc, long qty )
 
         obj->contents.emplace_back( *ammo );
         loc.remove_item();
+        ammo = nullptr;
```
Comment (mugling) [inline src/item.cpp]: You can just `return` here and skip any further branches

File: src/item.cpp [hunk 2]
```diff
@@ -4633,7 +4634,7 @@ bool item::reload( player &u, item_location loc, long qty )
         }
     }
 
-    if( ammo->charges == 0 ) {
+    if( ammo && ammo->charges == 0 ) {
```
Comment (mugling) [inline src/item.cpp]: ...which avoids the need for this test

Comment (tycho) [inline src/item.cpp]: Yeah, thought about that first. Then thought that perhaps it was desirable to have one return point. I can just turn it into a return...

**Comment by mugling** [comment]
Good find. I'd prefer it if we returned immediately after `remove_item` to help avoid this bug being reintroduced at a later date.

**Comment by tycho** [comment]
Done. For what it's worth, that was my original approach, but I figured people might want a single return point.

**Comment by mugling** [comment]
Thanks for the fix and your other recent PR's

**Comment by tycho** [comment]
@mugling You're welcome! I'm having fun poking at things, and happy to help.

**Pull request merged**